python/taichi/lang/struct.py

```diff
@@ -0,0 +1,570 @@
+        if len(args) == 1 and kwargs == {} and isinstance(args[0], dict):
+            self.entries = args[0]
+        elif len(args) == 0:
+            self.entries = kwargs
+        else:
+            raise TaichiSyntaxError(
+                "Custom structs need to be initialized using either dictionary or keyword arguments"
+            )
+        for k, v in self.entries.items():
+            if isinstance(v, (list, tuple)):
+                v = Matrix(v)
+            if isinstance(v, dict):
+                v = Struct(v)
+            self.entries[k] = v
+        self.register_members()
+        self.local_tensor_proxy = None
+        self.any_array_access = None
+
+    @property
+    def keys(self):
+        return list(self.entries.keys())
+
+    @property
+    def members(self):
+        return list(self.entries.values())
+
+    @property
+    def items(self):
+        return self.entries.items()
+
+    def register_members(self):
+        for k in self.keys:
+            setattr(Struct, k,
+                    property(
+                        Struct.make_getter(k),
+                        Struct.make_setter(k),
+                    ))
+
+    def __getitem__(self, key):
+        _taichi_skip_traceback = 1
+        ret = self.entries[key]
+        if isinstance(ret, SNodeHostAccess):
+            ret = ret.accessor.getter(*ret.key)
+        return ret
+
+    def __setitem__(self, key, value):
+        _taichi_skip_traceback = 1
+        if isinstance(self.entries[key], SNodeHostAccess):
+            self.entries[key].accessor.setter(value, *self.entries[key].key)
+        else:
+            if in_python_scope():
+                self.entries[key].set_entries(value)
+            else:
+                self.entries[key] = value
+
+    def set_entries(self, value):
+        if isinstance(value, dict):
+            value = Struct(value)
+        for k in self.keys:
+            self[k] = value[k]
+
+    @staticmethod
+    def make_getter(key):
+        def getter(self):
+            """Get an entry from custom struct by name."""
+            _taichi_skip_traceback = 1
+            return self[key]
+
+        return getter
+
+    @staticmethod
+    def make_setter(key):
+        @python_scope
+        def setter(self, value):
+            _taichi_skip_traceback = 1
+            self[key] = value
+
+        return setter
+
+    def element_wise_unary(self, foo):
+        _taichi_skip_traceback = 1
+        ret = self.empty_copy()
+        for k, v in self.items:
+            if isinstance(v, expr.Expr):
+                ret.entries[k] = foo(v)
+            else:
+                ret.entries[k] = v.element_wise_unary(foo)
+        return ret
+
+    def element_wise_binary(self, foo, other):
+        _taichi_skip_traceback = 1
+        ret = self.empty_copy()
+        if isinstance(other, (dict)):
+            other = Struct(other)
+        if isinstance(other, Struct):
+            assert self.entries.keys() == other.entries.keys(
```

Comment:
We use `assert` to verify whether the compiler itself is implemented correctly, and `raise` errors for user errors.